### ISSUE TITLE
V6 unable to handle authentication error through signInWithRedirect

### ISSUE THREAD

**Issue by pablovargasosorio** [aws-amplify/amplify-js#12651]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

Angular

### Amplify APIs

Authentication

### Amplify Categories

auth

### Environment information

<details>

```
  System:
    OS: macOS 14.1.1
    CPU: (8) arm64 Apple M1 Pro
    Memory: 42.28 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 18.18.2 - ~/.nvm/versions/node/v18.18.2/bin/node
    npm: 9.8.1 - ~/.nvm/versions/node/v18.18.2/bin/npm
  Browsers:
    Chrome: 119.0.6045.159
    Edge: 119.0.2151.72
    Safari: 17.1
  npmPackages:
    @angular-devkit/build-angular: ^16.2.8 => 16.2.8
    @angular/animations: ^16.2.11 => 16.2.11
    @angular/cdk: ^16.2.11 => 16.2.11
    @angular/cli: ^16.2.8 => 16.2.8
    @angular/common: ^16.2.11 => 16.2.11
    @angular/compiler: ^16.2.11 => 16.2.11
    @angular/compiler-cli: ^16.2.11 => 16.2.11
    @angular/core: ^16.2.11 => 16.2.11
    @angular/forms: ^16.2.11 => 16.2.11
    @angular/localize: ^16.2.11 => 16.2.11
    @angular/platform-browser: ^16.2.11 => 16.2.11
    @angular/platform-browser-dynamic: ^16.2.11 => 16.2.11
    @angular/router: ^16.2.11 => 16.2.11
    @auth0/angular-jwt: ^5.2.0 => 5.2.0
    @cypress/angular:  0.0.0-development
    @cypress/mount-utils:  0.0.0-development
    @cypress/react:  0.0.0-development
    @cypress/react18:  0.0.0-development
    @cypress/schematic: ^2.5.1 => 2.5.1
    @cypress/svelte:  0.0.0-development
    @cypress/vue:  0.0.0-development
    @cypress/vue2:  0.0.0-development
    @fortawesome/angular-fontawesome: ^0.13.0 => 0.13.0
    @fortawesome/fontawesome-free: ^6.4.2 => 6.4.2
    @fortawesome/fontawesome-svg-core: ^6.4.2 => 6.4.2
    @fortawesome/free-brands-svg-icons: ^6.4.2 => 6.4.2
    @fortawesome/free-regular-svg-icons: ^6.4.2 => 6.4.2
    @fortawesome/free-solid-svg-icons: ^6.4.2 => 6.4.2
    @ng-bootstrap/ng-bootstrap: ^15.1.2 => 15.1.2
    @ngrx/effects: ^16.3.0 => 16.3.0
    @ngrx/entity: ^16.3.0 => 16.3.0
    @ngrx/router-store: ^16.3.0 => 16.3.0
    @ngrx/store: ^16.3.0 => 16.3.0
    @ngrx/store-devtools: ^16.3.0 => 16.3.0
    @types/bootstrap: ^5.2.8 => 5.2.8
    @types/canvas-confetti: ^1.4.3 => 1.6.0
    @types/jasmine: ~3.8.0 => 3.8.2
    @types/lodash: ^4.14.182 => 4.14.192
    @types/node: ^12.11.1 => 12.20.55 (20.3.1, 16.18.40)
    @types/segment-analytics: ^0.0.34 => 0.0.34
    @types/textarea-caret: ^3.0.3 => 3.0.3
    @typescript-eslint/eslint-plugin: ^5.40.1 => 5.57.0
    @typescript-eslint/parser: ^5.40.1 => 5.57.0
    aws-amplify: ^6.0.5 => 6.0.5
    aws-amplify/adapter-core:  undefined ()
    aws-amplify/analytics:  undefined ()
    aws-amplify/analytics/kinesis:  undefined ()
    aws-amplify/analytics/kinesis-firehose:  undefined ()
    aws-amplify/analytics/personalize:  undefined ()
    aws-amplify/analytics/pinpoint:  undefined ()
    aws-amplify/api:  undefined ()
    aws-amplify/api/server:  undefined ()
    aws-amplify/auth:  undefined ()
    aws-amplify/auth/cognito:  undefined ()
    aws-amplify/auth/cognito/server:  undefined ()
    aws-amplify/auth/server:  undefined ()
    aws-amplify/datastore:  undefined ()
    aws-amplify/in-app-messaging:  undefined ()
    aws-amplify/in-app-messaging/pinpoint:  undefined ()
    aws-amplify/push-notifications:  undefined ()
    aws-amplify/push-notifications/pinpoint:  undefined ()
    aws-amplify/storage:  undefined ()
    aws-amplify/storage/s3:  undefined ()
    aws-amplify/storage/s3/server:  undefined ()
    aws-amplify/storage/server:  undefined ()
    aws-amplify/utils:  undefined ()
    bootstrap: ^5.3.2 => 5.3.2
    bootstrap-icons: ^1.11.1 => 1.11.1
    canvas-confetti: ^1.9.0 => 1.9.0
    chart.js: ^4.4.0 => 4.4.0
    chart.js-auto:  undefined ()
    chart.js-helpers:  undefined ()
    cypress: ^13.1.0 => 13.1.0
    cypress-localstorage-commands: ^2.2.1 => 2.2.4
    eslint: ^8.0.0 => 8.37.0
    eslint-config-prettier: ^8.5.0 => 8.8.0
    eslint-plugin-html: ^7.1.0 => 7.1.0
    eslint-plugin-tsdoc: ^0.2.14 => 0.2.17
    gmail-tester: ^1.3.8 => 1.3.8
    jasmine-core: ~3.8.0 => 3.8.0
    karma: ~6.3.0 => 6.3.20
    karma-chrome-launcher: ~3.1.0 => 3.1.1
    karma-coverage: ~2.0.3 => 2.0.3
    karma-coverage-coffee-example:  1.0.0
    karma-jasmine: ~4.0.0 => 4.0.2
    karma-jasmine-html-reporter: ~1.7.0 => 1.7.0
    karma-spec-reporter: ^0.0.36 => 0.0.36
    karma-verbose-reporter: 0.0.8 => 0.0.8
    lodash: ^4.17.21 => 4.17.21
    ng-circle-progress: ^1.7.1 => 1.7.1
    ng2-charts: ^5.0.3 => 5.0.3
    ng2-charts-schematics:  0.0.1
    ngx-cookie-service: ^16.1.0 => 16.1.0
    ngx-webstorage-service: ^5.0.0 => 5.0.0
    patternomaly: ^1.3.2 => 1.3.2
    prettier: 2.8.0 => 2.8.0
    replace-in-file: ^6.2.0 => 6.3.5
    rxjs: ^7.5.6 => 7.8.1 (6.6.7)
    rxjs/ajax:  undefined ()
    rxjs/fetch:  undefined ()
    rxjs/internal-compatibility:  undefined ()
    rxjs/operators:  undefined ()
    rxjs/testing:  undefined ()
    rxjs/webSocket:  undefined ()
    textarea-caret: ^3.1.0 => 3.1.0
    tslib: ^2.6.2 => 2.6.2 (2.6.1, 1.14.1)
    typescript: ~5.1.3 => 5.1.3
    zone.js: ~0.13.3 => 0.13.3
  npmGlobalPackages:
    @aws-amplify/cli: 12.8.2
    corepack: 0.19.0
    npm: 9.8.1
```

</details>

### Describe the bug

In aws-amplify/auth V5, after a login authentication error, I got redirected to _/redirect?error_description=...&error=unauthorized&state=......._
The Amplify library was automatically handling that and dispatching a **Hub** event called 'signIn_failure', so I was able to capture it and show the event.message which had the 'error_description'.

After upgrading to V6, I runned into the following problems:
- That Hub 'signIn_failure' event is deprecated, in favor of 'signInWithRedirect_failure', that's OK.
- Now that event doesn't have the 'message' property with the details of the error, it's been just removed from the event model.
- Anyway, the subscription that I had with _Hub.listen('auth', ..._ is not working anymore, because I just have a Unhandled exception JS error, then the app breaks.
- The exception is being raised from ZoneJS, that means I'm unable to catch it. So I only have a blank page with a JS error like this:
<img width="814" alt="Captura de pantalla 2023-11-29 a las 16 41 53" src="https://github.com/aws-amplify/amplify-js/assets/18540292/d66b5a4f-932f-41b8-9fcc-45c04a3aa4cb">

### Expected behavior

I need a way to handle the error when signInWithRedirect fails.

### Reproduction steps

1. Project is working with Amplify/auth V5.
2. Authentication error is handled through Hub 'signIn_failure' event listener.
3. Upgrade the project Amplify lib to V6.
4. Try to login with an account which raises an OAuth authentication failure. So the app is redirected to /redirect?error_description=... (as configured in Cognito)

### Code Snippet

Error handling snippet
```javascript
Hub.listen('auth', (data) => {
      const { payload } = data;
      if (payload.event === 'signInWithRedirect_failure') {
        // Legacy event was 'signIn_failure' (v5)
        // const msg = payload.data?.message || ''; DEPRECATED error message in amplify/auth v6
        const msg = 'There was an error signing in.';
        this.localStorage.addKey(
          'OAuthError',
          msg
        );
      }
});
```

My login button is executing this code:

```javascript
signInWithRedirect({ provider: 'Google' })
```

We jump to authentication dialog, then we come back to the URL /redirect/?error_description=...&error=unauthorized&state=.....
My code is just trying to initialize Amplify with:

```javascript
Amplify.configure(config);
```

Then it breaks. I tried with a try/catch like this:
```javascript
try {
  Amplify.configure(config);
} catch (e) { ...
```
It won't work because the exception is raised in Zone, I have no way to catch it...
And the error Hub listener above is not executed.

### Log output

<details>

```
[DEBUG] 52:54.682 Hub - Dispatching to core with  {event: 'configure', data: {…}}
core.mjs:26656 Angular is running in development mode.
config.service.ts:19 ERROR Error: Uncaught (in promise): OAuthSignInException: Email field is required; error=unauthorized
OAuthSignInException: Email field is required; error=unauthorized
    at handleFailure (signInWithRedirect.mjs:269:11)
    at signInWithRedirect.mjs:224:13
    at Generator.next (<anonymous>)
    at asyncGeneratorStep (asyncToGenerator.js:3:1)
    at _next (asyncToGenerator.js:22:1)
    at asyncToGenerator.js:27:1
    at new ZoneAwarePromise (zone.js:1411:21)
    at asyncToGenerator.js:19:1
    at _handleAuthResponse (signInWithRedirect.mjs:247:2)
    at handleAuthResponse (signInWithRedirect.mjs:218:34)
    at resolvePromise (zone.js:1193:31)
    at zone.js:1100:17
    at zone.js:1116:33
    at asyncGeneratorStep (asyncToGenerator.js:6:1)
    at _next (asyncToGenerator.js:22:1)
    at asyncToGenerator.js:27:1
    at new ZoneAwarePromise (zone.js:1411:21)
    at asyncToGenerator.js:19:1
    at _handleAuthResponse (signInWithRedirect.mjs:247:2)
    at handleAuthResponse (signInWithRedirect.mjs:218:34)
```

</details>

### Manual configuration

```
{
   "aws_cognito_region":"MY-REGION",
   "aws_project_region":"MY-REGION",
   "aws_user_pools_id":"MY-USER-POOLS-ID",
   "aws_user_pools_web_client_id":"MY-USER-POOLS-CLIENT-ID",
   "federationTarget":"COGNITO_USER_POOLS",
   "oauth":{
      "domain":"MY-DOMAIN",
      "redirectSignIn":"MY-HOST/redirect/",
      "redirectSignOut":"MY-HOST/logout/",
      "responseType":"code",
      "scope":[
         "email",
         "openid",
         "profile"
      ]
   }
}
```

**Comment by nadetastic** [comment]
Hi @pablovargasosorio thanks for opening this issue. Im currently taking a look at this and will follow up with you soon. In the meantime let me know if you have additional questions or comments.

**Comment by nadetastic** [comment]
Following up here @pablovargasosorio thanks for bring this to our attention. We've confirmed it as a bug and have identified a fix. Will provide another update soon.

**Comment by nadetastic** [comment]
Hi @pablovargasosorio a release containing a fix for this issue has been published on version `aws-amplify@6.0.6`. Could you give it a try and let me know if you have any issues?

**Comment by nadetastic** [comment]
@alexcracea are you calling in on a client side or server side?

**Comment by alexcracea** [comment]
BTW this error never happened with `aws-amplify/auth@6.0.1` and only reproduced after I updated

**Comment by nadetastic** [comment]
@alexcracea unfortunately I haven't been able to reproduce your issue with
```json
"aws-amplify": "6.0.5",
"next": "14.0.2"
```
Can you try deleting your node_modules and lock files and try reinstalling and building again?

If you still face an error after that, can you share a minimal sample app that produces this issue?

**Comment by pablovargasosorio** [comment]
Hey @nadetastic I've confirmed the Uncaught exception error has been fixed in v6.0.6
Thank you very much!

**Comment by nadetastic** [comment]
Hi @alexcracea following up here, let me know if you are still facing an issue, thanks!
